Application warning banner

mpmadhavig opened the pull request:
### Purpose
This PR will add a banner to the Application edit to maintain the below configs.

1. useClientIdAsSubClaimForAppTokens
2. omitUsernameInIntrospectionRespForAppTokens

The initial UI design is as follows.

https://github.com/user-attachments/assets/092abdbe-66a5-4813-98d4-1bdb359c6b4c

### Related Issues
- https://github.com/wso2/product-is/issues/20917

### Related PRs
- https://github.com/wso2/identity-apps/pull/6783

### Checklist

- [x] e2e cypress tests locally verified. (for internal contributers)
- [x] Manual test round performed and verified.
- [x] UX/UI review done on the final implementation.
- [x] Documentation provided. (Add links if there are any)
- [ ] Relevant backend changes deployed and verified
- [x] [Unit tests](/docs/testing/UNIT_TESTING.md) provided. (Add links if there are any)
- [x] [Integration tests](https://github.com/wso2/product-is/tree/master/modules/integration) provided. (Add links if there are any)

### Security checks
- [x] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines
- [x] Ran FindSecurityBugs plugin and verified report
- [x] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets

Achintha444 commented:
### Design comments

1. IMO it is not required the box title in the view more section.
2. IMO, it is not required to have the numbered list, only the check box is enough.
3. Content in the box, including the main button should be align with the content of the alert box.

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -57,13 +67,17 @@ import {
 import { ApplicationManagementUtils } from "../utils/application-management-utils";
 import { ApplicationTemplateManagementUtils } from "../utils/application-template-management-utils";
 import "./application-edit.scss";
+import { Typography } from "@mui/material";
```
do not directly import from material, use `https://wso2.github.io/oxygen-ui/react/?path=/docs/data-display-typography--overview` instead.

mpmadhavig replied:
Fixed with 025df58bd7452704342e396767d22883776319ca

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -129,6 +143,44 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+     * Assign loaded banner data into config states.
+     */
+    useEffect(() => {
+        if (applicationInboundConfig != undefined) {
```
```suggestion
        if (applicationInboundConfig) {
```

mpmadhavig replied:
Fixed with 025df58bd7452704342e396767d22883776319ca

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +558,354 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+     */
+    const resolveAlertBanner = (): ReactElement => {
+
+        const classes: any = classNames( { "application-outdated-alert-expanded-view": viewBannerDetails } );
```
refrain from using any.

Achintha444 replied:
I think we can replace `any` with `string` here.

mpmadhavig replied:
ce3f7f0210aea52dca4e1340286cc5ffcb701a43

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +558,354 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+     * @returns Alert banner.
+     */
+    const resolveAlertBanner = (): ReactElement => {
+
```
```suggestion
```

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +558,354 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+     *
+     * @returns Alert banner details.
+     */
+    const resolveBannerViewDetails = (): ReactElement => {
+
+        return (
```
```suggestion
    const resolveBannerViewDetails = (): ReactElement => {
```

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -33,21 +37,27 @@ import { ExtensionTemplateListInterface } from "@wso2is/admin.template-core.v1/m
 import { isFeatureEnabled } from "@wso2is/core/helpers";
 import { AlertLevels, IdentifiableComponentInterface } from "@wso2is/core/models";
 import { addAlert } from "@wso2is/core/store";
+import { Field, Forms } from "@wso2is/forms";
```
do not use `Forms` use `FinalForm` instead.
view this PR for more details, "https://github.com/wso2/identity-apps/pull/6722/files#diff-d94522cb3bcc2277d7af7c0f1c38e4e8d1d7659edb195199b995d045be70b235"

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +558,354 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+     *
+     * @returns Confirmation modal.
+     */
+    const confirmationModal = () => {
+
+        return (
```
```suggestion
    const confirmationModal = (): ReactElement => {
```

mpmadhavig replied:
Fixed with 025df58bd7452704342e396767d22883776319ca

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +558,354 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+     * Handles banner content update action which prepares data.
+     */
+    const handleBannerCheckBoxUpdate = (formDataValues: any) => {
+
```
```suggestion
```

mpmadhavig replied:
Fixed with 025df58bd7452704342e396767d22883776319ca

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +558,354 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+    /**
+     * Handles banner content update action which prepares data.
+     */
+    const handleBannerCheckBoxUpdate = (formDataValues: any) => {
```
refrain from using `any`

Achintha444 replied:
add return type

mpmadhavig replied:
ce3f7f0210aea52dca4e1340286cc5ffcb701a43

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +558,354 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+    const handleBannerCheckBoxUpdate = (formDataValues: any) => {
+
+        // Todo: check if we need to send all configs. Though the method is PUT, it works as a PATCH.
+        let values: any = { ...applicationInboundConfig };
```
refrain from using `any`

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +558,354 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+     * Handles the banner data update action.
+     */
+    const handleBannerCheckBoxUpdateConfirmation = async (): Promise<void> => {
+
```
```suggestion
```

mpmadhavig replied:
Fixed with 025df58bd7452704342e396767d22883776319ca

Achintha444 commented:
fix eslint issues.

mpmadhavig commented:
### Updated UI

1. Application edit section

<img width="1512" alt="Screenshot 2024-08-26 at 23 37 25" src="https://github.com/user-attachments/assets/ad2a717f-2993-4027-a69e-152953c1fc0c">

2. View Details view

<img width="1512" alt="Screenshot 2024-08-26 at 23 37 37" src="https://github.com/user-attachments/assets/0f513be6-34c1-4599-bf88-1ccf549eb997">

3. Application update confirmation modal action

<img width="1512" alt="Screenshot 2024-08-26 at 23 37 52" src="https://github.com/user-attachments/assets/99247977-b869-4751-8311-8c1a4f93f51f">

mpmadhavig commented:
> ### Design comments
> 
>     1. IMO it is not required the box title in the view more section.
> 
Removed with 007118f157fb1adbc5fd9ac54c862d7182289f42

>     2. IMO, it is not required to have the numbered list, only the check box is enough.
Removed with 007118f157fb1adbc5fd9ac54c862d7182289f42

>     3. Content in the box, including the main button should be align with the content of the alert box.
Fixed with 025df58bd7452704342e396767d22883776319ca

Achintha444 commented:
> > ### Design comments
> > ```
> > 1. IMO it is not required the box title in the view more section.
> > ```
> 
> Removed with [007118f](https://github.com/wso2/identity-apps/commit/007118f157fb1adbc5fd9ac54c862d7182289f42)
> 
> > ```
> > 2. IMO, it is not required to have the numbered list, only the check box is enough.
> > ```
> 
> Removed with [007118f](https://github.com/wso2/identity-apps/commit/007118f157fb1adbc5fd9ac54c862d7182289f42)
> 
> > ```
> > 3. Content in the box, including the main button should be align with the content of the alert box.
> > ```
> 
> Fixed with [025df58](https://github.com/wso2/identity-apps/commit/025df58bd7452704342e396767d22883776319ca)
> 
> <img alt="Screenshot 2024-08-27 at 21 15 13" width="1512" src="https://private-user-images.githubusercontent.com/47152272/361881503-39899744-3cf5-4e28-b04c-b9bb76ecb79b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ4MTMzMjMsIm5iZiI6MTcyNDgxMzAyMywicGF0aCI6Ii80NzE1MjI3Mi8zNjE4ODE1MDMtMzk4OTk3NDQtM2NmNS00ZTI4LWIwNGMtYjliYjc2ZWNiNzliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI4VDAyNDM0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxMjcxNGNiNmRhNDdjMTIwYzFhNTIwZjI3M2EwOTgwOGQyNDVkOTJkM2E3ZmY4Y2IzYjJiMjc1MGM5MjQzN2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.d-QTQ0F7ZJPqT8LRsMwV7fwevZ5BV-EmQts--oTb5p8">

### Design comments - Cont.

1. IMO the title above the each checkbox is not necessary, I think we are only rephrasing the same thing in two places. I think we can add a title to the box, which can summarise the both check boxes if necessary.

cc: @savindi7 @jeradrutnam WDYT ?

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +562,355 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+     *
+     * @returns Alert banner details.
+     */
+    const resolveBannerViewDetails = (): ReactElement => {
+        return (
```
```suggestion
    const resolveBannerViewDetails = (): ReactElement => {
```

mpmadhavig replied:
ce3f7f0210aea52dca4e1340286cc5ffcb701a43

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +562,355 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+     *
+     * @returns Confirmation modal.
+     */
+    const confirmationModal = () => {
+        return (
```
```suggestion
    const confirmationModal = (): ReactElement => {
```

Achintha444 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +562,355 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+    const handleBannerCheckBoxUpdateConfirmation = async (): Promise<void> => {
+        setBannerUpdateLoading(true);
+
+        return updateAuthProtocolConfig<OIDCDataInterface>(application?.id, formData, SupportedAuthProtocolTypes.OIDC)
```
why returning this function and the return type set as `void` ?

mpmadhavig replied:
ce3f7f0210aea52dca4e1340286cc5ffcb701a43

mpmadhavig commented:
### Updated UI designs

1. Application List View
<img width="1162" alt="Screenshot 2024-09-27 at 18 20 41" src="https://github.com/user-attachments/assets/4ebebe05-adad-4fec-8021-25644a23a2cf">

2. One outdated application edit view
<img width="1174" alt="Screenshot 2024-09-27 at 18 20 34" src="https://github.com/user-attachments/assets/dee791fb-d17d-45d7-a087-9fbaa99ea9fd">

3. Expanded banner view
<img width="1172" alt="Screenshot 2024-09-27 at 18 20 25" src="https://github.com/user-attachments/assets/00a630c4-92fe-496d-a2d7-1ca17099df5f">

pavinduLakshan commented on features/admin.core.v1/models/common.ts:
```diff
@@ -1,10 +1,19 @@
 /**
- * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com). All Rights Reserved.
+ * Copyright (c) 2024, WSO2 LLC. (https://www.wso2.com).
```
```suggestion
 * Copyright (c) 2023-2024, WSO2 LLC. (https://www.wso2.com).
```

pavinduLakshan commented on modules/react-components/src/components/page-header/page-header.tsx:
```diff
@@ -301,6 +306,9 @@ export const PageHeader: React.FunctionComponent<PageHeaderPropsInterface> = (
                                 )
                         )
                     }
+                    {
+                        alertBanner
+                    }
```
```suggestion
                    { alertBanner }
```

pavinduLakshan commented on modules/theme/src/themes/default/elements/label.overrides:
```diff
@@ -142,6 +142,17 @@
+
+.ui.label {
+    &.application-outdated-label {
+        color: #fd9200;
+        background-color: #fff4e5;
```
Can we use pre defined css variables for this?

cc: @savindi7

pavinduLakshan commented on modules/i18n/src/translations/en-US/portals/applications.ts:
```diff
@@ -1643,6 +1643,37 @@ export const applications: ApplicationsNS = {
+                                    instruction: "Introspection response for application access token will not include the username attribute."
+                                },
+                                useClientIdAsSubClaimOfAppTokens: {
+                                    instruction: "Application access token sub attribute will be client_id generated for an application."
```
```suggestion
                                    instruction: "The sub attribute of the application access token will be the client_id generated for the application."
```

pavinduLakshan commented on modules/i18n/src/translations/en-US/portals/applications.ts:
```diff
@@ -1643,6 +1643,37 @@ export const applications: ApplicationsNS = {
+                        versions: {
+                            version100: {
+                                removeUsernameFromIntrospectionRespForAppTokens: {
+                                    instruction: "Introspection response for application access token will not include the username attribute."
```
```suggestion
                                    instruction: "The introspection response for an application access token will not include the username attribute."
```

pavinduLakshan commented on modules/i18n/src/translations/en-US/portals/applications.ts:
```diff
@@ -1643,6 +1643,37 @@ export const applications: ApplicationsNS = {
+                    confirmationModal: {
+                        header: "Have you done the relevant changes?",
+                        message: "Proceeding the action without making relevant change will cause the client application behavior break.",
+                        content: "This action is irreversible and will permanently update the application.",
```
```suggestion
                        content: "This action is irreversible and will result in a permanent update to the application.",
```

pavinduLakshan commented on modules/i18n/src/translations/en-US/portals/applications.ts:
```diff
@@ -1643,6 +1643,37 @@ export const applications: ApplicationsNS = {
+                    documentationHint: "More Details",
+                    confirmationModal: {
+                        header: "Have you done the relevant changes?",
+                        message: "Proceeding the action without making relevant change will cause the client application behavior break.",
```
```suggestion
                        message: "Proceeding without making the necessary changes will cause the client application's behavior to break.",
```

pavinduLakshan reviewed (commented):
shall we get the UI text reviewed from the doc team?

cc: @himeshsiriwardana

pavinduLakshan commented on features/admin.applications.v1/constants/application-management.ts:
```diff
@@ -41,6 +41,11 @@ export class ApplicationManagementConstants {
+    /**
+     * Application latest version.
+     */
+    public static readonly LATEST_VERSION: string = "v1.0.0";
```
Is hardcoding this here intentional?

pavinduLakshan replied:
Let's add a comment

mpmadhavig replied:
0bd8ceec802ce087b6d7e10d520ad8d787c6e5f1

pavinduLakshan commented on features/admin.applications.v1/utils/application-management-utils.ts:
```diff
@@ -104,6 +104,28 @@ export class ApplicationManagementUtils {
             });
     }
 
+    public static getIfAppIsOutdated(applicationVersion: string, grantTypes?: string[]): boolean {
```
```suggestion
    public static isApplicationOutdated(applicationVersion: string, grantTypes?: string[]): boolean {
```

mpmadhavig replied:
0bd8ceec802ce087b6d7e10d520ad8d787c6e5f1

mpmadhavig commented:
Due to the comment in https://github.com/wso2-enterprise/iam-product-management/issues/295#issuecomment-2385261432, the label will not be displayed.

pavinduLakshan commented on features/admin.applications.v1/utils/application-management-utils.ts:
```diff
@@ -104,6 +104,28 @@ export class ApplicationManagementUtils {
 
+    public static isApplicationOutdated(applicationVersion: string, grantTypes?: string[]): boolean {
+
+        if (applicationVersion != undefined
+                && grantTypes.includes(ApplicationManagementConstants.CLIENT_CREDENTIALS_GRANT)) {
```
```suggestion
        if (
             applicationVersion != undefined
             && grantTypes.includes(ApplicationManagementConstants.CLIENT_CREDENTIALS_GRANT)
        ) {
```

himeshsiriwardana commented:
**Application Update Required.**

We have updated the behavior of application access tokens as follows::

- The `sub` attribute of an application access token now returns the `client_id` generated for the application, instead of the `userid` of the application owner.
- The introspection responses for application access tokens no longer return the `username` attribute.

You may click the button below to update the application’s behavior. Once updated, be sure to update any usages of these attributes accordingly. [Learn more]

mpmadhavig commented:
IS banner view:
<img width="1123" alt="Screenshot 2024-10-08 at 15 16 10" src="https://github.com/user-attachments/assets/5f026e8b-c18b-4d8b-a229-61dc61de2433">

savindi7 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -129,6 +149,25 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+    const [ formData, setFormdata ] = useState<ApplicationInterface>(undefined);
+
+    useEffect(() => {
+        if (application != undefined && applicationInboundConfigs != undefined) {
```
```suggestion
        if (application && applicationInboundConfigs) {
```

savindi7 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +545,297 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+                            severity="warning"
+                            action={
+                                (
+                                    <Box display="flex">
```
let's not use inline styles and add these in the stylesheets.

savindi7 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +545,297 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+                                            className="ignore-once-button"
+                                            onClick={ () => setDisplayBanner(false) }>
+                                            <Icon
+                                                link={ true }
```
```suggestion
                                                link
```

savindi7 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -506,6 +545,297 @@ const ApplicationEditPage: FunctionComponent<ApplicationEditPageInterface> = (
+                                            <Icon
+                                                link={ true }
+                                                onClick={ () => setDisplayBanner(false) }
+                                                className=""
```
```suggestion
```

savindi7 commented on features/admin.applications.v1/pages/application-edit.tsx:
```diff
@@ -33,21 +42,29 @@ import { ExtensionTemplateListInterface } from "@wso2is/admin.template-core.v1/m
 import { RouteComponentProps } from "react-router";
 import { Dispatch } from "redux";
-import { Label } from "semantic-ui-react";
+import { Icon, Label } from "semantic-ui-react";
```
Let's use the oxygen components here

savindi7 commented on modules/theme/src/themes/default/elements/label.overrides:
```diff
@@ -142,6 +142,17 @@
+*******************************/
+
+.ui.label {
+    &.application-outdated-label {
```
Are we using these styles anywhere?

mpmadhavig commented:
Will make the necessary changes via the next phase of this effort.
